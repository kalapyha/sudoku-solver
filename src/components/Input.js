import React from 'react';

function Input() {
	return (
		<div className="d-flex">
			<input
				type="text"
				placeholder="Create your own Sudoku. ex: ...1...2...7..4...23.."
			/>
			<button className="btn btn-primary">Create</button>
		</div>
	);
}

export default Input;
