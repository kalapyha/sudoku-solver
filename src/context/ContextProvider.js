import React, { useState, useContext } from 'react';
import {
	defaultBaord_9by9,
	emptyBaord_9by9,
	solveSudoku,
} from '../utils/sudoku_solver';

const LoadingContext = React.createContext();
const BoardToSolveContext = React.createContext();
const ResultBoardContext = React.createContext();
const SudokuSolveContext = React.createContext();

export function useLoading() {
	return useContext(LoadingContext);
}

export function useBoardToSolve() {
	return useContext(BoardToSolveContext);
}

export function useResultBoard() {
	return useContext(ResultBoardContext);
}

export function useSolveBoard() {
	return useContext(SudokuSolveContext);
}

export function ContextProvider({ children }) {
	const [isLoading, setisLoading] = useState(false);
	const [boardToSolve, setboardToSolve] = useState(defaultBaord_9by9);
	const [resultBoard, setResultBoard] = useState(emptyBaord_9by9);

	function handleSolve() {
		setisLoading(true);
		const { board, time } = solveSudoku(boardToSolve);
		setResultBoard(board);

		setisLoading(false);
		console.log(`Time spent to solve Sudoku: ${time}ms`);
	}
	return (
		<LoadingContext.Provider value={isLoading}>
			<BoardToSolveContext.Provider value={boardToSolve}>
				<ResultBoardContext.Provider value={resultBoard}>
					<SudokuSolveContext.Provider value={handleSolve}>
						{children}
					</SudokuSolveContext.Provider>
				</ResultBoardContext.Provider>
			</BoardToSolveContext.Provider>
		</LoadingContext.Provider>
	);
}
