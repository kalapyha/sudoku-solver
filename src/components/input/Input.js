import React, { useState } from 'react';
import { useGlobalContext } from '../../context/ContextProvider';
import { isValidBoard, createBoard } from '../../utils/helpers';

function Input() {
	const { updateBoard, clearBoard, showErrorMessage } = useGlobalContext();

	const [input, setInput] = useState('');

	const handleClick = () => {
		if (!isValidBoard(createBoard(input))) {
			showErrorMessage();
		} else {
			updateBoard(createBoard(input));
			setInput('');
			clearBoard();
		}
	};

	const handleChange = (e) => {
		const inputValue = e.target.value.replace(/[^1-9.]/g, '').slice(0, 81);
		setInput(inputValue);
	};

	return (
		<div className="d-flex">
			<input
				data-testid="input"
				type="text"
				placeholder="Create your own Sudoku. ex: ...1...2...7..4...23.."
				value={input}
				onChange={handleChange}
			/>
			<button
				data-testid="input-btn"
				className="btn btn-primary"
				disabled={input.length === 81 ? false : true}
				onClick={handleClick}
			>
				Create
			</button>
		</div>
	);
}

export default Input;
