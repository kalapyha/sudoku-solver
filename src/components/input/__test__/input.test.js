import React from 'react';
import { AppProvider } from '../../../context/ContextProvider';
import { render, fireEvent } from '@testing-library/react';
import TestRenderer from 'react-test-renderer';
import Input from '../Input';

const setup = () => {
	const utils = render(
		<AppProvider>
			<Input />
		</AppProvider>
	);
	const input = utils.getByTestId('input');
	const inputBtn = utils.getByTestId('input-btn');
	return {
		input,
		inputBtn,
		...utils,
	};
};

describe('<Input />', () => {
	it('it should ignore everything except 1-9 and .', () => {
		const { input } = setup();
		fireEvent.change(input, { target: { value: '0 test @#$%^&*() 42 ..' } });
		expect(input.value).toBe('42..');
	});

	it('submit button should be disabled', () => {
		const { inputBtn } = setup();
		expect(inputBtn).toBeDisabled();
	});

	it('matches snapshot', () => {
		const tree = TestRenderer.create(
			<AppProvider>
				<Input />
			</AppProvider>
		).toJSON();
		expect(tree).toMatchSnapshot();
	});
});
