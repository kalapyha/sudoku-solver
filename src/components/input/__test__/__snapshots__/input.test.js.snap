// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> matches snapshot 1`] = `
<div
  className="d-flex"
>
  <input
    data-testid="input"
    onChange={[Function]}
    placeholder="Create your own Sudoku. ex: ...1...2...7..4...23.."
    type="text"
    value=""
  />
  <button
    className="btn btn-primary"
    data-testid="input-btn"
    disabled={true}
    onClick={[Function]}
  >
    Create
  </button>
</div>
`;
